keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,330 @@
+- Define how `ImageDecryptSecret` can be used by configuration yaml(s) of the Pod (or Deployments)
+- Define how `ImageDecryptSecret` can be integrated into the service accounts
+- Define the Image Authorization process to prevent unauthorized access to the cached encrypted images. 
+
+
+### Non-Goals
+
+- Kubernetes should be able to decrypt the images. However, in a typical workflow kubernetes has no role to play to encrypt the images. This is similar to how kubernetes plays a role in downloading and using the images instead of building and uploading container images to the registry.  
+
+### User Stories
+
+- As a cluster user, I want to create a secret that carries the private keys required for my encrypted images
+- As a cluster user, I want to run the encrypted container images using the private keys carried by ImageDecryptSecret
+- As a cluster operator, I want to add the ImageDecryptSecret to the service account
+- As a cluster user, I want to run the encrypted images using the private keys from the service account
+- As an application developer, I want to encrypt my container images and be able to run them securely using kubernetes
+- As an application developer, I want to protect the content of my container images such that only me (as an application developer) and the execution runtime can read them. Any other third party, such as container registry, should **NOT** be able to read the content of my container image.
+
+## Proposal
+
+The initial design includes:
+
+- `ImageDecryptSecret` API resource definition
+- `ImageDecryptSecret` pod field for specifying the ImageDecryptSecret the pod should be run with
+- Kubelet implementation for fetching & interpreting the ImageDecryptSecret
+- CRI API & implementation for passing along the ImageDecryptSecret
+
+### API
+
+`ImageDecryptSecret` is a new cluster-scoped Secret
+
+
+The private key is selected by the pod by specifying the ImageDecryptSecret in the PodSpec. Once the pod is
+scheduled, the ImageDecryptSecret cannot be changed.
+
+_(This is a simplified declaration, syntactic details will be covered in the API PR review)_
+
+```go
+// Container represents a single container that is expected to be run on the host.
+type Container struct {
+  <snip>
+  // ImageDecryptSecrets is an optional list of references to secrets in the same namespace to use for decrypting any of the images used by this PodSpec.
+  // If specified, these secrets will be passed to individual puller implementations for them to use.
+  // +optional
+  ImageDecryptSecrets []LocalObjectReference
+  </snip>
+}
+```
+[Reference Implementation](https://github.com/harche/kubernetes/blob/8a0ef9898a55110d7e41f8e1a846c66cfc2fa691/pkg/apis/core/types.go#L1915) of the `Container` struct
+
+```go
+const (
+  <snip>
+  // SecretTypeDecryptKeys defines the type for the decrypt secrets
+  SecretTypeDecryptKey SecretType = "kubernetes.io/decryptionkey"
```

Comment:
Looking at the other `SecretType`* constants in this file, they tend to document/describe/constraint the format of the contents of the secret. What is the format of `SecretTypeDecryptKey` secrets?

Reply:
An `ImageDecryptSecret` can hold one or more private key(s) with their corresponding optional passwords. Secret value is set [here](https://github.com/harche/kubernetes/blob/pr_branch/staging/src/k8s.io/kubectl/pkg/generate/versioned/secret_for_decrypt_image.go#L87) which is just a JSON serialized value of [this](https://github.com/harche/kubernetes/blob/pr_branch/staging/src/k8s.io/kubectl/pkg/generate/versioned/secret_for_decrypt_image.go#L130) the struct `DecryptConfigEntry`.